// @hashflow/taker-js/dist/rest.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.createRestApi = void 0;
const axios_1 = __importDefault(require("axios"));
const http_1 = __importDefault(require("http"));
const https_1 = __importDefault(require("https"));
const URL_BY_ENV = {
    production: 'https://api.hashflow.com',
    staging: 'https://api-staging.hashflow.com',
    development: undefined,
};
function createRestApi(env, authKey) {
    let baseURL = URL_BY_ENV[env];
    if (!baseURL) {
        const customURL = process.env.HASHFLOW_API_URL;
        if (!customURL) {
            throw new Error(`No API URL found for environment ${env}`);
        }
        else {
            baseURL = customURL;
        }
    }
    return axios_1.default.create({
        baseURL,
        headers: { Authorization: authKey },
        httpAgent: new http_1.default.Agent({ keepAlive: true }),
        httpsAgent: new https_1.default.Agent({ keepAlive: true }),
    });
}
exports.createRestApi = createRestApi;
