// @hashflow/contracts-evm/dist/types/factories/contracts/interfaces/external/ILayerZeroRelayer__factory.js
"use strict";
/* Autogenerated file. Do not edit manually. */
/* tslint:disable */
/* eslint-disable */
Object.defineProperty(exports, "__esModule", { value: true });
exports.ILayerZeroRelayer__factory = void 0;
const ethers_1 = require("ethers");
const _abi = [
    {
        inputs: [
            {
                internalType: "uint16",
                name: "_dstChainId",
                type: "uint16",
            },
            {
                internalType: "uint16",
                name: "_outboundProofType",
                type: "uint16",
            },
            {
                internalType: "address",
                name: "_userApplication",
                type: "address",
            },
            {
                internalType: "uint256",
                name: "_payloadSize",
                type: "uint256",
            },
            {
                internalType: "bytes",
                name: "_adapterParams",
                type: "bytes",
            },
        ],
        name: "getPrice",
        outputs: [
            {
                internalType: "uint256",
                name: "price",
                type: "uint256",
            },
        ],
        stateMutability: "view",
        type: "function",
    },
    {
        inputs: [
            {
                internalType: "address",
                name: "_address",
                type: "address",
            },
        ],
        name: "isApproved",
        outputs: [
            {
                internalType: "bool",
                name: "approved",
                type: "bool",
            },
        ],
        stateMutability: "view",
        type: "function",
    },
    {
        inputs: [
            {
                internalType: "uint16",
                name: "_dstChainId",
                type: "uint16",
            },
            {
                internalType: "uint16",
                name: "_outboundProofType",
                type: "uint16",
            },
            {
                internalType: "bytes",
                name: "_adapterParams",
                type: "bytes",
            },
        ],
        name: "notifyRelayer",
        outputs: [],
        stateMutability: "nonpayable",
        type: "function",
    },
];
class ILayerZeroRelayer__factory {
    static createInterface() {
        return new ethers_1.Interface(_abi);
    }
    static connect(address, runner) {
        return new ethers_1.Contract(address, _abi, runner);
    }
}
exports.ILayerZeroRelayer__factory = ILayerZeroRelayer__factory;
ILayerZeroRelayer__factory.abi = _abi;
