// @hashflow/contracts-evm/dist/types/factories/@openzeppelin/contracts/utils/ShortStrings__factory.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.ShortStrings__factory = void 0;
/* Autogenerated file. Do not edit manually. */
/* tslint:disable */
/* eslint-disable */
const ethers_1 = require("ethers");
const _abi = [
    {
        inputs: [],
        name: "InvalidShortString",
        type: "error",
    },
    {
        inputs: [
            {
                internalType: "string",
                name: "str",
                type: "string",
            },
        ],
        name: "StringTooLong",
        type: "error",
    },
];
const _bytecode = "0x60566037600b82828239805160001a607314602a57634e487b7160e01b600052600060045260246000fd5b30600052607381538281f3fe73000000000000000000000000000000000000000030146080604052600080fdfea264697066735822122078fea987bc3cc0f0d7c6d303b111bc63f934adb87c6354bc0fec2e3e2ccbb94f64736f6c63430008120033";
const isSuperArgs = (xs) => xs.length > 1;
class ShortStrings__factory extends ethers_1.ContractFactory {
    constructor(...args) {
        if (isSuperArgs(args)) {
            super(...args);
        }
        else {
            super(_abi, _bytecode, args[0]);
        }
    }
    getDeployTransaction(overrides) {
        return super.getDeployTransaction(overrides || {});
    }
    deploy(overrides) {
        return super.deploy(overrides || {});
    }
    connect(runner) {
        return super.connect(runner);
    }
    static createInterface() {
        return new ethers_1.Interface(_abi);
    }
    static connect(address, runner) {
        return new ethers_1.Contract(address, _abi, runner);
    }
}
exports.ShortStrings__factory = ShortStrings__factory;
ShortStrings__factory.bytecode = _bytecode;
ShortStrings__factory.abi = _abi;
