// @hashflow/taker-js/dist/types/common.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.CHAIN_IDS_BY_TYPE = exports.SOLANA_CHAIN_IDS = exports.SOLANA_DEVNET_CHAIN_ID = exports.SOLANA_MAINNET_CHAIN_ID = exports.RFQType = void 0;
const utils_1 = require("@hashflow/contracts-evm/dist/src/utils");
var RFQType;
(function (RFQType) {
    RFQType[RFQType["RFQT"] = 0] = "RFQT";
    RFQType[RFQType["RFQM"] = 1] = "RFQM";
})(RFQType = exports.RFQType || (exports.RFQType = {}));
exports.SOLANA_MAINNET_CHAIN_ID = 1000001;
exports.SOLANA_DEVNET_CHAIN_ID = 1000002;
exports.SOLANA_CHAIN_IDS = [
    exports.SOLANA_MAINNET_CHAIN_ID,
    exports.SOLANA_DEVNET_CHAIN_ID,
];
exports.CHAIN_IDS_BY_TYPE = {
    evm: [
        utils_1.ETHEREUM.chainId,
        utils_1.GOERLI.chainId,
        utils_1.OPTIMISM.chainId,
        utils_1.POLYGON.chainId,
        utils_1.MUMBAI.chainId,
        utils_1.BNB.chainId,
        utils_1.BNB_TESTNET.chainId,
        utils_1.ARBITRUM.chainId,
        utils_1.AVALANCHE.chainId,
        utils_1.BASE.chainId,
        utils_1.SEPOLIA.chainId,
    ],
    solana: [exports.SOLANA_MAINNET_CHAIN_ID, exports.SOLANA_DEVNET_CHAIN_ID],
};
