// @hashflow/sdk/dist/contracts/hashverse/factories/MEMint__factory.js
"use strict";
/* Autogenerated file. Do not edit manually. */
/* tslint:disable */
/* eslint-disable */
Object.defineProperty(exports, "__esModule", { value: true });
exports.MEMint__factory = void 0;
const ethers_1 = require("ethers");
const _abi = [
    {
        inputs: [
            {
                internalType: "bytes32",
                name: "preAllocatedMerkleRoot",
                type: "bytes32",
            },
            {
                internalType: "bytes32",
                name: "publicMerkleRoot",
                type: "bytes32",
            },
            {
                internalType: "address",
                name: "renovaItem",
                type: "address",
            },
            {
                internalType: "uint256",
                name: "maxPublicMints",
                type: "uint256",
            },
        ],
        stateMutability: "nonpayable",
        type: "constructor",
    },
    {
        anonymous: false,
        inputs: [
            {
                indexed: false,
                internalType: "address",
                name: "player",
                type: "address",
            },
        ],
        name: "Mint",
        type: "event",
    },
    {
        inputs: [
            {
                internalType: "uint256[]",
                name: "hashverseItemIds",
                type: "uint256[]",
            },
            {
                internalType: "bytes32[]",
                name: "proof",
                type: "bytes32[]",
            },
        ],
        name: "mint",
        outputs: [],
        stateMutability: "nonpayable",
        type: "function",
    },
    {
        inputs: [],
        name: "totalMinted",
        outputs: [
            {
                internalType: "uint256",
                name: "",
                type: "uint256",
            },
        ],
        stateMutability: "view",
        type: "function",
    },
];
class MEMint__factory {
    static createInterface() {
        return new ethers_1.Interface(_abi);
    }
    static connect(address, runner) {
        return new ethers_1.Contract(address, _abi, runner);
    }
}
exports.MEMint__factory = MEMint__factory;
MEMint__factory.abi = _abi;
