// @hashflow/protocol/dist/typechain-types/factories/contracts/interfaces/external/ILayerZeroOracle__factory.js
"use strict";
/* Autogenerated file. Do not edit manually. */
/* tslint:disable */
/* eslint-disable */
Object.defineProperty(exports, "__esModule", { value: true });
exports.ILayerZeroOracle__factory = void 0;
const ethers_1 = require("ethers");
const _abi = [
    {
        inputs: [
            {
                internalType: "uint16",
                name: "_dstChainId",
                type: "uint16",
            },
            {
                internalType: "uint16",
                name: "_outboundProofType",
                type: "uint16",
            },
        ],
        name: "getPrice",
        outputs: [
            {
                internalType: "uint256",
                name: "price",
                type: "uint256",
            },
        ],
        stateMutability: "view",
        type: "function",
    },
    {
        inputs: [
            {
                internalType: "address",
                name: "_address",
                type: "address",
            },
        ],
        name: "isApproved",
        outputs: [
            {
                internalType: "bool",
                name: "approved",
                type: "bool",
            },
        ],
        stateMutability: "view",
        type: "function",
    },
    {
        inputs: [
            {
                internalType: "uint16",
                name: "_dstChainId",
                type: "uint16",
            },
            {
                internalType: "uint16",
                name: "_outboundProofType",
                type: "uint16",
            },
            {
                internalType: "uint64",
                name: "_outboundBlockConfirmations",
                type: "uint64",
            },
        ],
        name: "notifyOracle",
        outputs: [],
        stateMutability: "nonpayable",
        type: "function",
    },
];
class ILayerZeroOracle__factory {
    static createInterface() {
        return new ethers_1.utils.Interface(_abi);
    }
    static connect(address, signerOrProvider) {
        return new ethers_1.Contract(address, _abi, signerOrProvider);
    }
}
exports.ILayerZeroOracle__factory = ILayerZeroOracle__factory;
ILayerZeroOracle__factory.abi = _abi;
