// @hashflow/protocol/dist/typechain-types/factories/contracts/interfaces/external/ILayerZeroNonceContract__factory.js
"use strict";
/* Autogenerated file. Do not edit manually. */
/* tslint:disable */
/* eslint-disable */
Object.defineProperty(exports, "__esModule", { value: true });
exports.ILayerZeroNonceContract__factory = void 0;
const ethers_1 = require("ethers");
const _abi = [
    {
        inputs: [
            {
                internalType: "uint16",
                name: "dstChainId",
                type: "uint16",
            },
            {
                internalType: "bytes",
                name: "path",
                type: "bytes",
            },
        ],
        name: "outboundNonce",
        outputs: [
            {
                internalType: "uint64",
                name: "",
                type: "uint64",
            },
        ],
        stateMutability: "view",
        type: "function",
    },
];
class ILayerZeroNonceContract__factory {
    static createInterface() {
        return new ethers_1.utils.Interface(_abi);
    }
    static connect(address, signerOrProvider) {
        return new ethers_1.Contract(address, _abi, signerOrProvider);
    }
}
exports.ILayerZeroNonceContract__factory = ILayerZeroNonceContract__factory;
ILayerZeroNonceContract__factory.abi = _abi;
