// @hashflow/protocol/dist/typechain-types/factories/contracts/interfaces/external/ILayerZeroMessagingLibrary__factory.js
"use strict";
/* Autogenerated file. Do not edit manually. */
/* tslint:disable */
/* eslint-disable */
Object.defineProperty(exports, "__esModule", { value: true });
exports.ILayerZeroMessagingLibrary__factory = void 0;
const ethers_1 = require("ethers");
const _abi = [
    {
        inputs: [
            {
                internalType: "uint16",
                name: "_chainId",
                type: "uint16",
            },
            {
                internalType: "address",
                name: "_userApplication",
                type: "address",
            },
            {
                internalType: "bytes",
                name: "_payload",
                type: "bytes",
            },
            {
                internalType: "bool",
                name: "_payInZRO",
                type: "bool",
            },
            {
                internalType: "bytes",
                name: "_adapterParam",
                type: "bytes",
            },
        ],
        name: "estimateFees",
        outputs: [
            {
                internalType: "uint256",
                name: "nativeFee",
                type: "uint256",
            },
            {
                internalType: "uint256",
                name: "zroFee",
                type: "uint256",
            },
        ],
        stateMutability: "view",
        type: "function",
    },
    {
        inputs: [
            {
                internalType: "uint16",
                name: "_chainId",
                type: "uint16",
            },
            {
                internalType: "address",
                name: "_userApplication",
                type: "address",
            },
            {
                internalType: "uint256",
                name: "_configType",
                type: "uint256",
            },
        ],
        name: "getConfig",
        outputs: [
            {
                internalType: "bytes",
                name: "",
                type: "bytes",
            },
        ],
        stateMutability: "view",
        type: "function",
    },
    {
        inputs: [
            {
                internalType: "address",
                name: "_userApplication",
                type: "address",
            },
            {
                internalType: "uint64",
                name: "_lastNonce",
                type: "uint64",
            },
            {
                internalType: "uint16",
                name: "_chainId",
                type: "uint16",
            },
            {
                internalType: "bytes",
                name: "_destination",
                type: "bytes",
            },
            {
                internalType: "bytes",
                name: "_payload",
                type: "bytes",
            },
            {
                internalType: "address payable",
                name: "refundAddress",
                type: "address",
            },
            {
                internalType: "address",
                name: "_zroPaymentAddress",
                type: "address",
            },
            {
                internalType: "bytes",
                name: "_adapterParams",
                type: "bytes",
            },
        ],
        name: "send",
        outputs: [],
        stateMutability: "payable",
        type: "function",
    },
    {
        inputs: [
            {
                internalType: "uint16",
                name: "_chainId",
                type: "uint16",
            },
            {
                internalType: "address",
                name: "_userApplication",
                type: "address",
            },
            {
                internalType: "uint256",
                name: "_configType",
                type: "uint256",
            },
            {
                internalType: "bytes",
                name: "_config",
                type: "bytes",
            },
        ],
        name: "setConfig",
        outputs: [],
        stateMutability: "nonpayable",
        type: "function",
    },
];
class ILayerZeroMessagingLibrary__factory {
    static createInterface() {
        return new ethers_1.utils.Interface(_abi);
    }
    static connect(address, signerOrProvider) {
        return new ethers_1.Contract(address, _abi, signerOrProvider);
    }
}
exports.ILayerZeroMessagingLibrary__factory = ILayerZeroMessagingLibrary__factory;
ILayerZeroMessagingLibrary__factory.abi = _abi;
