// @hashflow/protocol/dist/typechain-types/factories/contracts/interfaces/external/IOpynController__factory.js
"use strict";
/* Autogenerated file. Do not edit manually. */
/* tslint:disable */
/* eslint-disable */
Object.defineProperty(exports, "__esModule", { value: true });
exports.IOpynController__factory = void 0;
const ethers_1 = require("ethers");
const _abi = [
    {
        inputs: [
            {
                components: [
                    {
                        internalType: "enum IOpynActions.ActionType",
                        name: "actionType",
                        type: "uint8",
                    },
                    {
                        internalType: "address",
                        name: "owner",
                        type: "address",
                    },
                    {
                        internalType: "address",
                        name: "secondAddress",
                        type: "address",
                    },
                    {
                        internalType: "address",
                        name: "asset",
                        type: "address",
                    },
                    {
                        internalType: "uint256",
                        name: "vaultId",
                        type: "uint256",
                    },
                    {
                        internalType: "uint256",
                        name: "amount",
                        type: "uint256",
                    },
                    {
                        internalType: "uint256",
                        name: "index",
                        type: "uint256",
                    },
                    {
                        internalType: "bytes",
                        name: "data",
                        type: "bytes",
                    },
                ],
                internalType: "struct IOpynActions.ActionArgs[]",
                name: "_actions",
                type: "tuple[]",
            },
        ],
        name: "operate",
        outputs: [],
        stateMutability: "nonpayable",
        type: "function",
    },
];
class IOpynController__factory {
    static createInterface() {
        return new ethers_1.utils.Interface(_abi);
    }
    static connect(address, signerOrProvider) {
        return new ethers_1.Contract(address, _abi, signerOrProvider);
    }
}
exports.IOpynController__factory = IOpynController__factory;
IOpynController__factory.abi = _abi;
