// @hashflow/sdk/dist/contracts/protocol/index.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || (function () {
    var ownKeys = function(o) {
        ownKeys = Object.getOwnPropertyNames || function (o) {
            var ar = [];
            for (var k in o) if (Object.prototype.hasOwnProperty.call(o, k)) ar[ar.length] = k;
            return ar;
        };
        return ownKeys(o);
    };
    return function (mod) {
        if (mod && mod.__esModule) return mod;
        var result = {};
        if (mod != null) for (var k = ownKeys(mod), i = 0; i < k.length; i++) if (k[i] !== "default") __createBinding(result, mod, k[i]);
        __setModuleDefault(result, mod);
        return result;
    };
})();
Object.defineProperty(exports, "__esModule", { value: true });
exports.IWormhole__factory = exports.IWETH__factory = exports.ILayerZeroNonceContract__factory = exports.ILayerZeroEndpoint__factory = exports.IHashflowWormholeMessenger__factory = exports.IHashflowRouter__factory = exports.IHashflowPool__factory = exports.IHashflowLayerZeroMessenger__factory = exports.IHashflowFactory__factory = exports.IHashflowAavePortal__factory = exports.IERC20__factory = exports.ERC20__factory = exports.factories = void 0;
exports.factories = __importStar(require("./factories"));
var ERC20__factory_1 = require("./factories/ERC20__factory");
Object.defineProperty(exports, "ERC20__factory", { enumerable: true, get: function () { return ERC20__factory_1.ERC20__factory; } });
var IERC20__factory_1 = require("./factories/IERC20__factory");
Object.defineProperty(exports, "IERC20__factory", { enumerable: true, get: function () { return IERC20__factory_1.IERC20__factory; } });
var IHashflowAavePortal__factory_1 = require("./factories/IHashflowAavePortal__factory");
Object.defineProperty(exports, "IHashflowAavePortal__factory", { enumerable: true, get: function () { return IHashflowAavePortal__factory_1.IHashflowAavePortal__factory; } });
var IHashflowFactory__factory_1 = require("./factories/IHashflowFactory__factory");
Object.defineProperty(exports, "IHashflowFactory__factory", { enumerable: true, get: function () { return IHashflowFactory__factory_1.IHashflowFactory__factory; } });
var IHashflowLayerZeroMessenger__factory_1 = require("./factories/IHashflowLayerZeroMessenger__factory");
Object.defineProperty(exports, "IHashflowLayerZeroMessenger__factory", { enumerable: true, get: function () { return IHashflowLayerZeroMessenger__factory_1.IHashflowLayerZeroMessenger__factory; } });
var IHashflowPool__factory_1 = require("./factories/IHashflowPool__factory");
Object.defineProperty(exports, "IHashflowPool__factory", { enumerable: true, get: function () { return IHashflowPool__factory_1.IHashflowPool__factory; } });
var IHashflowRouter__factory_1 = require("./factories/IHashflowRouter__factory");
Object.defineProperty(exports, "IHashflowRouter__factory", { enumerable: true, get: function () { return IHashflowRouter__factory_1.IHashflowRouter__factory; } });
var IHashflowWormholeMessenger__factory_1 = require("./factories/IHashflowWormholeMessenger__factory");
Object.defineProperty(exports, "IHashflowWormholeMessenger__factory", { enumerable: true, get: function () { return IHashflowWormholeMessenger__factory_1.IHashflowWormholeMessenger__factory; } });
var ILayerZeroEndpoint__factory_1 = require("./factories/ILayerZeroEndpoint__factory");
Object.defineProperty(exports, "ILayerZeroEndpoint__factory", { enumerable: true, get: function () { return ILayerZeroEndpoint__factory_1.ILayerZeroEndpoint__factory; } });
var ILayerZeroNonceContract__factory_1 = require("./factories/ILayerZeroNonceContract__factory");
Object.defineProperty(exports, "ILayerZeroNonceContract__factory", { enumerable: true, get: function () { return ILayerZeroNonceContract__factory_1.ILayerZeroNonceContract__factory; } });
var IWETH__factory_1 = require("./factories/IWETH__factory");
Object.defineProperty(exports, "IWETH__factory", { enumerable: true, get: function () { return IWETH__factory_1.IWETH__factory; } });
var IWormhole__factory_1 = require("./factories/IWormhole__factory");
Object.defineProperty(exports, "IWormhole__factory", { enumerable: true, get: function () { return IWormhole__factory_1.IWormhole__factory; } });
