// @hashflow/contracts-evm/dist/types/factories/contracts/interfaces/index.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || function (mod) {
    if (mod && mod.__esModule) return mod;
    var result = {};
    if (mod != null) for (var k in mod) if (k !== "default" && Object.prototype.hasOwnProperty.call(mod, k)) __createBinding(result, mod, k);
    __setModuleDefault(result, mod);
    return result;
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.IHashflowRouter__factory = exports.IHashflowPool__factory = exports.IHashflowFactory__factory = exports.xchain = exports.xapp = exports.external = void 0;
/* Autogenerated file. Do not edit manually. */
/* tslint:disable */
/* eslint-disable */
exports.external = __importStar(require("./external"));
exports.xapp = __importStar(require("./xapp"));
exports.xchain = __importStar(require("./xchain"));
var IHashflowFactory__factory_1 = require("./IHashflowFactory__factory");
Object.defineProperty(exports, "IHashflowFactory__factory", { enumerable: true, get: function () { return IHashflowFactory__factory_1.IHashflowFactory__factory; } });
var IHashflowPool__factory_1 = require("./IHashflowPool__factory");
Object.defineProperty(exports, "IHashflowPool__factory", { enumerable: true, get: function () { return IHashflowPool__factory_1.IHashflowPool__factory; } });
var IHashflowRouter__factory_1 = require("./IHashflowRouter__factory");
Object.defineProperty(exports, "IHashflowRouter__factory", { enumerable: true, get: function () { return IHashflowRouter__factory_1.IHashflowRouter__factory; } });
