// @hashflow/sdk/dist/legacy/v1_0/contracts/protocol/index.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || (function () {
    var ownKeys = function(o) {
        ownKeys = Object.getOwnPropertyNames || function (o) {
            var ar = [];
            for (var k in o) if (Object.prototype.hasOwnProperty.call(o, k)) ar[ar.length] = k;
            return ar;
        };
        return ownKeys(o);
    };
    return function (mod) {
        if (mod && mod.__esModule) return mod;
        var result = {};
        if (mod != null) for (var k = ownKeys(mod), i = 0; i < k.length; i++) if (k[i] !== "default") __createBinding(result, mod, k[i]);
        __setModuleDefault(result, mod);
        return result;
    };
})();
Object.defineProperty(exports, "__esModule", { value: true });
exports.TestWeth__factory = exports.TestToken2__factory = exports.TestToken1__factory = exports.IWeth__factory = exports.ILayerZeroNonceContract__factory = exports.ILayerZeroEndpoint__factory = exports.IHashflowXChainUA__factory = exports.IHashflowTracker__factory = exports.IHashflowRouter__factory = exports.IHashflowPublicPool__factory = exports.IHashflowPrivatePool__factory = exports.IHashflowPool__factory = exports.IHashflowGovernance__factory = exports.IHashflowFactory__factory = exports.IHashflowERC20__factory = exports.IERC20__factory = exports.factories = void 0;
exports.factories = __importStar(require("./factories"));
var IERC20__factory_1 = require("./factories/IERC20__factory");
Object.defineProperty(exports, "IERC20__factory", { enumerable: true, get: function () { return IERC20__factory_1.IERC20__factory; } });
var IHashflowERC20__factory_1 = require("./factories/IHashflowERC20__factory");
Object.defineProperty(exports, "IHashflowERC20__factory", { enumerable: true, get: function () { return IHashflowERC20__factory_1.IHashflowERC20__factory; } });
var IHashflowFactory__factory_1 = require("./factories/IHashflowFactory__factory");
Object.defineProperty(exports, "IHashflowFactory__factory", { enumerable: true, get: function () { return IHashflowFactory__factory_1.IHashflowFactory__factory; } });
var IHashflowGovernance__factory_1 = require("./factories/IHashflowGovernance__factory");
Object.defineProperty(exports, "IHashflowGovernance__factory", { enumerable: true, get: function () { return IHashflowGovernance__factory_1.IHashflowGovernance__factory; } });
var IHashflowPool__factory_1 = require("./factories/IHashflowPool__factory");
Object.defineProperty(exports, "IHashflowPool__factory", { enumerable: true, get: function () { return IHashflowPool__factory_1.IHashflowPool__factory; } });
var IHashflowPrivatePool__factory_1 = require("./factories/IHashflowPrivatePool__factory");
Object.defineProperty(exports, "IHashflowPrivatePool__factory", { enumerable: true, get: function () { return IHashflowPrivatePool__factory_1.IHashflowPrivatePool__factory; } });
var IHashflowPublicPool__factory_1 = require("./factories/IHashflowPublicPool__factory");
Object.defineProperty(exports, "IHashflowPublicPool__factory", { enumerable: true, get: function () { return IHashflowPublicPool__factory_1.IHashflowPublicPool__factory; } });
var IHashflowRouter__factory_1 = require("./factories/IHashflowRouter__factory");
Object.defineProperty(exports, "IHashflowRouter__factory", { enumerable: true, get: function () { return IHashflowRouter__factory_1.IHashflowRouter__factory; } });
var IHashflowTracker__factory_1 = require("./factories/IHashflowTracker__factory");
Object.defineProperty(exports, "IHashflowTracker__factory", { enumerable: true, get: function () { return IHashflowTracker__factory_1.IHashflowTracker__factory; } });
var IHashflowXChainUA__factory_1 = require("./factories/IHashflowXChainUA__factory");
Object.defineProperty(exports, "IHashflowXChainUA__factory", { enumerable: true, get: function () { return IHashflowXChainUA__factory_1.IHashflowXChainUA__factory; } });
var ILayerZeroEndpoint__factory_1 = require("./factories/ILayerZeroEndpoint__factory");
Object.defineProperty(exports, "ILayerZeroEndpoint__factory", { enumerable: true, get: function () { return ILayerZeroEndpoint__factory_1.ILayerZeroEndpoint__factory; } });
var ILayerZeroNonceContract__factory_1 = require("./factories/ILayerZeroNonceContract__factory");
Object.defineProperty(exports, "ILayerZeroNonceContract__factory", { enumerable: true, get: function () { return ILayerZeroNonceContract__factory_1.ILayerZeroNonceContract__factory; } });
var IWeth__factory_1 = require("./factories/IWeth__factory");
Object.defineProperty(exports, "IWeth__factory", { enumerable: true, get: function () { return IWeth__factory_1.IWeth__factory; } });
var TestToken1__factory_1 = require("./factories/TestToken1__factory");
Object.defineProperty(exports, "TestToken1__factory", { enumerable: true, get: function () { return TestToken1__factory_1.TestToken1__factory; } });
var TestToken2__factory_1 = require("./factories/TestToken2__factory");
Object.defineProperty(exports, "TestToken2__factory", { enumerable: true, get: function () { return TestToken2__factory_1.TestToken2__factory; } });
var TestWeth__factory_1 = require("./factories/TestWeth__factory");
Object.defineProperty(exports, "TestWeth__factory", { enumerable: true, get: function () { return TestWeth__factory_1.TestWeth__factory; } });
