// @hashflow/sdk/dist/index.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || (function () {
    var ownKeys = function(o) {
        ownKeys = Object.getOwnPropertyNames || function (o) {
            var ar = [];
            for (var k in o) if (Object.prototype.hasOwnProperty.call(o, k)) ar[ar.length] = k;
            return ar;
        };
        return ownKeys(o);
    };
    return function (mod) {
        if (mod && mod.__esModule) return mod;
        var result = {};
        if (mod != null) for (var k = ownKeys(mod), i = 0; i < k.length; i++) if (k[i] !== "default") __createBinding(result, mod, k[i]);
        __setModuleDefault(result, mod);
        return result;
    };
})();
var __exportStar = (this && this.__exportStar) || function(m, exports) {
    for (var p in m) if (p !== "default" && !Object.prototype.hasOwnProperty.call(exports, p)) __createBinding(exports, m, p);
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.XChainProtocol = exports.interfaces = exports.Hashflow = exports.IHashflowWormholeMessenger__factory = exports.IHashflowRouter__factory = exports.IHashflowPool__factory = exports.IHashflowFactory__factory = exports.IMultiStakingVault__factory = exports.contractAddressesByNetwork = void 0;
var Contracts_1 = require("./Contracts");
Object.defineProperty(exports, "contractAddressesByNetwork", { enumerable: true, get: function () { return Contracts_1.contractAddressesByNetwork; } });
var IMultiStakingVault__factory_1 = require("./contracts/hashverse/factories/IMultiStakingVault__factory");
Object.defineProperty(exports, "IMultiStakingVault__factory", { enumerable: true, get: function () { return IMultiStakingVault__factory_1.IMultiStakingVault__factory; } });
var IHashflowFactory__factory_1 = require("./contracts/protocol/factories/IHashflowFactory__factory");
Object.defineProperty(exports, "IHashflowFactory__factory", { enumerable: true, get: function () { return IHashflowFactory__factory_1.IHashflowFactory__factory; } });
var IHashflowPool__factory_1 = require("./contracts/protocol/factories/IHashflowPool__factory");
Object.defineProperty(exports, "IHashflowPool__factory", { enumerable: true, get: function () { return IHashflowPool__factory_1.IHashflowPool__factory; } });
var IHashflowRouter__factory_1 = require("./contracts/protocol/factories/IHashflowRouter__factory");
Object.defineProperty(exports, "IHashflowRouter__factory", { enumerable: true, get: function () { return IHashflowRouter__factory_1.IHashflowRouter__factory; } });
var IHashflowWormholeMessenger__factory_1 = require("./contracts/protocol/factories/IHashflowWormholeMessenger__factory");
Object.defineProperty(exports, "IHashflowWormholeMessenger__factory", { enumerable: true, get: function () { return IHashflowWormholeMessenger__factory_1.IHashflowWormholeMessenger__factory; } });
var Hashflow_1 = require("./Hashflow");
Object.defineProperty(exports, "Hashflow", { enumerable: true, get: function () { return Hashflow_1.Hashflow; } });
exports.interfaces = __importStar(require("./Interfaces"));
__exportStar(require("./lib/Constants"), exports);
__exportStar(require("./lib/Helpers"), exports);
__exportStar(require("./types"), exports);
var protocol_1 = require("@hashflow/protocol");
Object.defineProperty(exports, "XChainProtocol", { enumerable: true, get: function () { return protocol_1.XChainProtocol; } });
