// @hashflow/sdk/dist/legacy/v0_1/lib/Constants.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.contractAddressesByNetwork = exports.layerZeroChainIdByNetwork = exports.layerZeroEndpointByNetwork = exports.QUOTE_TYPEHASH = exports.ADDRESSES = exports.ZERO_ADDRESS = exports.SUBTRACT_GAS_LIMIT = void 0;
const Contracts_1 = require("../Contracts");
Object.defineProperty(exports, "contractAddressesByNetwork", { enumerable: true, get: function () { return Contracts_1.contractAddressesByNetwork; } });
const types_1 = require("../types");
exports.SUBTRACT_GAS_LIMIT = 100000;
exports.ZERO_ADDRESS = '0x0000000000000000000000000000000000000000';
exports.ADDRESSES = {
    BETH: '0xEeeeeEeeeEeEeeEeEeEeeEEEeeeeEeeeeeeeEEeE', //Balancer eth
    KOVAN: {
        WETH: '0xd0a1e359811322d97991e03f863a0c30c2cf029c',
        MARGIN: '0x5934807cC0654d46755eBd2848840b616256C6Ef',
    },
    ROPSTEN: {
        WETH: '0xc778417e063141139fce010982780140aa0cd5ab',
        MARGIN: ' 0x3C325EeBB64495665F5376930d30151C1075bFD8',
    },
    MAINNET: {
        WETH: '0xC02aaA39b223FE8D0A0e5C4F27eAD9083C756Cc2',
        MARGIN: '0x8c7c60d766951c5c570bbb7065c993070061b795',
    },
    MUMBAI: {
        WETH: '0x9c3C9283D3e44854697Cd22D3Faa240Cfb032889',
    },
};
exports.QUOTE_TYPEHASH = '0xee85c4f11aac87fb24a4a24b8add5dfcd9a4829992e61d9e7fdc704a35f4d5ae';
exports.layerZeroEndpointByNetwork = {
    [types_1.Networks.MAINNET]: '0x66A71Dcef29A0fFBDBE3c6a460a3B5BC225Cd675',
    [types_1.Networks.BSC]: '0x3c2269811836af69497E5F486A85D7316753cf62',
    [types_1.Networks.AVALANCHE]: '0x3c2269811836af69497E5F486A85D7316753cf62',
    [types_1.Networks.POLYGON]: '0x3c2269811836af69497E5F486A85D7316753cf62',
    [types_1.Networks.ARBITRUM]: '0x3c2269811836af69497E5F486A85D7316753cf62',
    [types_1.Networks.OPTIMISM]: '0x3c2269811836af69497E5F486A85D7316753cf62',
    [types_1.Networks.RINKEBY]: '0x79a63d6d8BBD5c6dfc774dA79bCcD948EAcb53FA',
    [types_1.Networks.BSCTESTNET]: '0x6Fcb97553D41516Cb228ac03FdC8B9a0a9df04A1',
    [types_1.Networks.MUMBAI]: '0xf69186dfBa60DdB133E91E9A4B5673624293d8F8',
};
exports.layerZeroChainIdByNetwork = {
    [types_1.Networks.MAINNET]: 101,
    [types_1.Networks.BSC]: 102,
    [types_1.Networks.AVALANCHE]: 106,
    [types_1.Networks.POLYGON]: 109,
    [types_1.Networks.ARBITRUM]: 110,
    [types_1.Networks.OPTIMISM]: 111,
    [types_1.Networks.RINKEBY]: 10001,
    [types_1.Networks.BSCTESTNET]: 10002,
    [types_1.Networks.MUMBAI]: 10009,
};
