// @hashflow/protocol/dist/typechain-types/factories/contracts/interfaces/IHashflowFactory__factory.js
"use strict";
/* Autogenerated file. Do not edit manually. */
/* tslint:disable */
/* eslint-disable */
Object.defineProperty(exports, "__esModule", { value: true });
exports.IHashflowFactory__factory = void 0;
const ethers_1 = require("ethers");
const _abi = [
    {
        anonymous: false,
        inputs: [
            {
                indexed: false,
                internalType: "address",
                name: "pool",
                type: "address",
            },
            {
                indexed: false,
                internalType: "address",
                name: "operations",
                type: "address",
            },
            {
                indexed: false,
                internalType: "bool",
                name: "privatePool",
                type: "bool",
            },
        ],
        name: "CreatePool",
        type: "event",
    },
    {
        anonymous: false,
        inputs: [
            {
                indexed: false,
                internalType: "address",
                name: "governance",
                type: "address",
            },
            {
                indexed: false,
                internalType: "address",
                name: "prevGovernance",
                type: "address",
            },
        ],
        name: "UpdateGovernance",
        type: "event",
    },
    {
        anonymous: false,
        inputs: [
            {
                indexed: false,
                internalType: "address",
                name: "hTokenImpl",
                type: "address",
            },
            {
                indexed: false,
                internalType: "address",
                name: "prevHTokenImpl",
                type: "address",
            },
        ],
        name: "UpdateHTokenImplementation",
        type: "event",
    },
    {
        anonymous: false,
        inputs: [
            {
                indexed: false,
                internalType: "address",
                name: "privatePoolImpl",
                type: "address",
            },
            {
                indexed: false,
                internalType: "address",
                name: "prevPrivatePoolImpl",
                type: "address",
            },
        ],
        name: "UpdatePrivatePoolImplementation",
        type: "event",
    },
    {
        anonymous: false,
        inputs: [
            {
                indexed: false,
                internalType: "address",
                name: "publicPoolImpl",
                type: "address",
            },
            {
                indexed: false,
                internalType: "address",
                name: "prevPublicPoolImpl",
                type: "address",
            },
        ],
        name: "UpdatePublicPoolImplementation",
        type: "event",
    },
    {
        inputs: [
            {
                internalType: "string",
                name: "name",
                type: "string",
            },
            {
                internalType: "string",
                name: "symbol",
                type: "string",
            },
            {
                internalType: "address",
                name: "signer",
                type: "address",
            },
            {
                internalType: "bool",
                name: "isPrivate",
                type: "bool",
            },
        ],
        name: "createPool",
        outputs: [],
        stateMutability: "nonpayable",
        type: "function",
    },
    {
        inputs: [
            {
                internalType: "address",
                name: "governance",
                type: "address",
            },
        ],
        name: "updateGovernance",
        outputs: [],
        stateMutability: "nonpayable",
        type: "function",
    },
    {
        inputs: [
            {
                internalType: "address",
                name: "hTokenImpl",
                type: "address",
            },
        ],
        name: "updateHTokenImpl",
        outputs: [],
        stateMutability: "nonpayable",
        type: "function",
    },
    {
        inputs: [
            {
                internalType: "address",
                name: "privatePoolImpl",
                type: "address",
            },
        ],
        name: "updatePrivatePoolImpl",
        outputs: [],
        stateMutability: "nonpayable",
        type: "function",
    },
    {
        inputs: [
            {
                internalType: "address",
                name: "publicPoolImpl",
                type: "address",
            },
        ],
        name: "updatePublicPoolImpl",
        outputs: [],
        stateMutability: "nonpayable",
        type: "function",
    },
];
class IHashflowFactory__factory {
    static createInterface() {
        return new ethers_1.utils.Interface(_abi);
    }
    static connect(address, signerOrProvider) {
        return new ethers_1.Contract(address, _abi, signerOrProvider);
    }
}
exports.IHashflowFactory__factory = IHashflowFactory__factory;
IHashflowFactory__factory.abi = _abi;
