// @hashflow/protocol/dist/typechain-types/factories/contracts/interfaces/IHashflowPublicPoolFactoryZkSync__factory.js
"use strict";
/* Autogenerated file. Do not edit manually. */
/* tslint:disable */
/* eslint-disable */
Object.defineProperty(exports, "__esModule", { value: true });
exports.IHashflowPublicPoolFactoryZkSync__factory = void 0;
const ethers_1 = require("ethers");
const _abi = [
    {
        inputs: [
            {
                internalType: "string",
                name: "name",
                type: "string",
            },
            {
                internalType: "string",
                name: "symbol",
                type: "string",
            },
            {
                internalType: "address",
                name: "signer",
                type: "address",
            },
            {
                internalType: "address",
                name: "operations",
                type: "address",
            },
            {
                internalType: "address",
                name: "governance",
                type: "address",
            },
            {
                internalType: "address",
                name: "router",
                type: "address",
            },
        ],
        name: "createPool",
        outputs: [
            {
                internalType: "address",
                name: "",
                type: "address",
            },
        ],
        stateMutability: "nonpayable",
        type: "function",
    },
];
class IHashflowPublicPoolFactoryZkSync__factory {
    static createInterface() {
        return new ethers_1.utils.Interface(_abi);
    }
    static connect(address, signerOrProvider) {
        return new ethers_1.Contract(address, _abi, signerOrProvider);
    }
}
exports.IHashflowPublicPoolFactoryZkSync__factory = IHashflowPublicPoolFactoryZkSync__factory;
IHashflowPublicPoolFactoryZkSync__factory.abi = _abi;
