// @hashflow/protocol/dist/typechain-types/factories/contracts/interfaces/index.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    Object.defineProperty(o, k2, { enumerable: true, get: function() { return m[k]; } });
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || function (mod) {
    if (mod && mod.__esModule) return mod;
    var result = {};
    if (mod != null) for (var k in mod) if (k !== "default" && Object.prototype.hasOwnProperty.call(mod, k)) __createBinding(result, mod, k);
    __setModuleDefault(result, mod);
    return result;
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.IHashflowXChainUA__factory = exports.IHashflowRouter__factory = exports.IHashflowPublicPoolFactoryZkSync__factory = exports.IHashflowPublicPool__factory = exports.IHashflowPrivatePoolFactoryZkSync__factory = exports.IHashflowPrivatePool__factory = exports.IHashflowPool__factory = exports.IHashflowGovernance__factory = exports.IHashflowFactoryZkSync__factory = exports.IHashflowFactory__factory = exports.IHashflowERC20__factory = exports.external = void 0;
/* Autogenerated file. Do not edit manually. */
/* tslint:disable */
/* eslint-disable */
exports.external = __importStar(require("./external"));
var IHashflowERC20__factory_1 = require("./IHashflowERC20__factory");
Object.defineProperty(exports, "IHashflowERC20__factory", { enumerable: true, get: function () { return IHashflowERC20__factory_1.IHashflowERC20__factory; } });
var IHashflowFactory__factory_1 = require("./IHashflowFactory__factory");
Object.defineProperty(exports, "IHashflowFactory__factory", { enumerable: true, get: function () { return IHashflowFactory__factory_1.IHashflowFactory__factory; } });
var IHashflowFactoryZkSync__factory_1 = require("./IHashflowFactoryZkSync__factory");
Object.defineProperty(exports, "IHashflowFactoryZkSync__factory", { enumerable: true, get: function () { return IHashflowFactoryZkSync__factory_1.IHashflowFactoryZkSync__factory; } });
var IHashflowGovernance__factory_1 = require("./IHashflowGovernance__factory");
Object.defineProperty(exports, "IHashflowGovernance__factory", { enumerable: true, get: function () { return IHashflowGovernance__factory_1.IHashflowGovernance__factory; } });
var IHashflowPool__factory_1 = require("./IHashflowPool__factory");
Object.defineProperty(exports, "IHashflowPool__factory", { enumerable: true, get: function () { return IHashflowPool__factory_1.IHashflowPool__factory; } });
var IHashflowPrivatePool__factory_1 = require("./IHashflowPrivatePool__factory");
Object.defineProperty(exports, "IHashflowPrivatePool__factory", { enumerable: true, get: function () { return IHashflowPrivatePool__factory_1.IHashflowPrivatePool__factory; } });
var IHashflowPrivatePoolFactoryZkSync__factory_1 = require("./IHashflowPrivatePoolFactoryZkSync__factory");
Object.defineProperty(exports, "IHashflowPrivatePoolFactoryZkSync__factory", { enumerable: true, get: function () { return IHashflowPrivatePoolFactoryZkSync__factory_1.IHashflowPrivatePoolFactoryZkSync__factory; } });
var IHashflowPublicPool__factory_1 = require("./IHashflowPublicPool__factory");
Object.defineProperty(exports, "IHashflowPublicPool__factory", { enumerable: true, get: function () { return IHashflowPublicPool__factory_1.IHashflowPublicPool__factory; } });
var IHashflowPublicPoolFactoryZkSync__factory_1 = require("./IHashflowPublicPoolFactoryZkSync__factory");
Object.defineProperty(exports, "IHashflowPublicPoolFactoryZkSync__factory", { enumerable: true, get: function () { return IHashflowPublicPoolFactoryZkSync__factory_1.IHashflowPublicPoolFactoryZkSync__factory; } });
var IHashflowRouter__factory_1 = require("./IHashflowRouter__factory");
Object.defineProperty(exports, "IHashflowRouter__factory", { enumerable: true, get: function () { return IHashflowRouter__factory_1.IHashflowRouter__factory; } });
var IHashflowXChainUA__factory_1 = require("./IHashflowXChainUA__factory");
Object.defineProperty(exports, "IHashflowXChainUA__factory", { enumerable: true, get: function () { return IHashflowXChainUA__factory_1.IHashflowXChainUA__factory; } });
