// @bgd-labs/aave-address-book/dist/AaveV3Mantle.js
"use strict";
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);

// src/ts/AaveV3Mantle.ts
var AaveV3Mantle_exports = {};
__export(AaveV3Mantle_exports, {
  AAVE_PROTOCOL_DATA_PROVIDER: () => AAVE_PROTOCOL_DATA_PROVIDER,
  ACL_ADMIN: () => ACL_ADMIN,
  ACL_MANAGER: () => ACL_MANAGER,
  ASSETS: () => ASSETS,
  CHAIN_ID: () => CHAIN_ID,
  COLLECTOR: () => COLLECTOR,
  CONFIG_ENGINE: () => CONFIG_ENGINE,
  DEFAULT_A_TOKEN_IMPL_REV_1: () => DEFAULT_A_TOKEN_IMPL_REV_1,
  DEFAULT_INCENTIVES_CONTROLLER: () => DEFAULT_INCENTIVES_CONTROLLER,
  DEFAULT_VARIABLE_DEBT_TOKEN_IMPL_REV_1: () => DEFAULT_VARIABLE_DEBT_TOKEN_IMPL_REV_1,
  DUST_BIN: () => DUST_BIN,
  EMISSION_MANAGER: () => EMISSION_MANAGER,
  EXTERNAL_LIBRARIES: () => EXTERNAL_LIBRARIES,
  E_MODES: () => E_MODES,
  L2_ENCODER: () => L2_ENCODER,
  ORACLE: () => ORACLE,
  POOL: () => POOL,
  POOL_ADDRESSES_PROVIDER: () => POOL_ADDRESSES_PROVIDER,
  POOL_ADDRESSES_PROVIDER_REGISTRY: () => POOL_ADDRESSES_PROVIDER_REGISTRY,
  POOL_CONFIGURATOR: () => POOL_CONFIGURATOR,
  POOL_CONFIGURATOR_IMPL: () => POOL_CONFIGURATOR_IMPL,
  POOL_IMPL: () => POOL_IMPL,
  RISK_STEWARD: () => RISK_STEWARD,
  STATA_FACTORY: () => STATA_FACTORY,
  UI_INCENTIVE_DATA_PROVIDER: () => UI_INCENTIVE_DATA_PROVIDER,
  UI_POOL_DATA_PROVIDER: () => UI_POOL_DATA_PROVIDER,
  WALLET_BALANCE_PROVIDER: () => WALLET_BALANCE_PROVIDER,
  WETH_GATEWAY: () => WETH_GATEWAY
});
module.exports = __toCommonJS(AaveV3Mantle_exports);
var POOL_ADDRESSES_PROVIDER = "0x2aB3580a805fB10CbAd567212C70e26C1B6769eC";
var POOL = "0x2e770EF8AbdEcA83D9310E2d3B3c2FdfFF5fd85A";
var POOL_CONFIGURATOR = "0x4fDE7000ac23033B0FD420098B8a83A9Ff52c33C";
var ORACLE = "0x6c23bAF050ec192afc0B967a93b83e6c5405df43";
var ACL_ADMIN = "0x70884634D0098782592111A2A6B8d223be31CB7b";
var ACL_MANAGER = "0x3e652E97ff339B73421f824F5b03d75b62F1Fb51";
var AAVE_PROTOCOL_DATA_PROVIDER = "0x9138E2cAdFEB23AFFdc0419F2912CaB8F135dba9";
var POOL_IMPL = "0x589750BA8aF186cE5B55391B0b7148cAD43a1619";
var POOL_CONFIGURATOR_IMPL = "0xBeC519531F0E78BcDdB295242fA4EC5251B38574";
var DEFAULT_INCENTIVES_CONTROLLER = "0xe875feF12DF97D763038F0Eac53962Ca36249eA6";
var EMISSION_MANAGER = "0x9D6aB154437A28CF202954e1176449c75e6D6c4B";
var COLLECTOR = "0x29B8Edc8a5158e8eBD7Dea3473517AB079260a0b";
var DEFAULT_A_TOKEN_IMPL_REV_1 = "0xd8247C7926841337E8adf9fD9E87026992C45062";
var DEFAULT_VARIABLE_DEBT_TOKEN_IMPL_REV_1 = "0x294daB91f3539f5a74Ed995a59235eE62A14A345";
var CONFIG_ENGINE = "0xea9989569Cf2D860597b4d649c9A963c15ab84fa";
var POOL_ADDRESSES_PROVIDER_REGISTRY = "0xe892E40C92c2E4D281Be59b2E6300F271d824E75";
var UI_INCENTIVE_DATA_PROVIDER = "0x01d678F1bbE148C96e7501F1Ac41661904F84F61";
var UI_POOL_DATA_PROVIDER = "0x4F6f44325828D2A40724A0a966F33d75cD1DF7c1";
var WALLET_BALANCE_PROVIDER = "0x4758213271BFdC72224A7a8742dC865fC97756e1";
var WETH_GATEWAY = "0xde35f7711295Dfd0CD8bd94d46f65b8Acb182520";
var L2_ENCODER = "0x4A49351ED179b23af08523Af98a8ca95C712b9d9";
var STATA_FACTORY = "0x7cCC8a3DF66a2cDEa6c0629412378752Db5014EA";
var RISK_STEWARD = "0xCCCcf90C363Eaf09dD89DD5330C1287Ff6a945Ee";
var DUST_BIN = "0x7485FB1552BB107fFc923f41E548e8e3FFea5E45";
var CHAIN_ID = 5e3;
var ASSETS = {};
var E_MODES = {};
var EXTERNAL_LIBRARIES = {};
// Annotate the CommonJS export names for ESM import in node:
0 && (module.exports = {
  AAVE_PROTOCOL_DATA_PROVIDER,
  ACL_ADMIN,
  ACL_MANAGER,
  ASSETS,
  CHAIN_ID,
  COLLECTOR,
  CONFIG_ENGINE,
  DEFAULT_A_TOKEN_IMPL_REV_1,
  DEFAULT_INCENTIVES_CONTROLLER,
  DEFAULT_VARIABLE_DEBT_TOKEN_IMPL_REV_1,
  DUST_BIN,
  EMISSION_MANAGER,
  EXTERNAL_LIBRARIES,
  E_MODES,
  L2_ENCODER,
  ORACLE,
  POOL,
  POOL_ADDRESSES_PROVIDER,
  POOL_ADDRESSES_PROVIDER_REGISTRY,
  POOL_CONFIGURATOR,
  POOL_CONFIGURATOR_IMPL,
  POOL_IMPL,
  RISK_STEWARD,
  STATA_FACTORY,
  UI_INCENTIVE_DATA_PROVIDER,
  UI_POOL_DATA_PROVIDER,
  WALLET_BALANCE_PROVIDER,
  WETH_GATEWAY
});
