// @bgd-labs/aave-address-book/dist/AaveV3Sonic.js
"use strict";
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);

// src/ts/AaveV3Sonic.ts
var AaveV3Sonic_exports = {};
__export(AaveV3Sonic_exports, {
  AAVE_PROTOCOL_DATA_PROVIDER: () => AAVE_PROTOCOL_DATA_PROVIDER,
  ACL_ADMIN: () => ACL_ADMIN,
  ACL_MANAGER: () => ACL_MANAGER,
  ASSETS: () => ASSETS,
  CHAIN_ID: () => CHAIN_ID,
  COLLECTOR: () => COLLECTOR,
  CONFIG_ENGINE: () => CONFIG_ENGINE,
  DEFAULT_A_TOKEN_IMPL_REV_1: () => DEFAULT_A_TOKEN_IMPL_REV_1,
  DEFAULT_INCENTIVES_CONTROLLER: () => DEFAULT_INCENTIVES_CONTROLLER,
  DEFAULT_VARIABLE_DEBT_TOKEN_IMPL_REV_1: () => DEFAULT_VARIABLE_DEBT_TOKEN_IMPL_REV_1,
  DUST_BIN: () => DUST_BIN,
  EMISSION_MANAGER: () => EMISSION_MANAGER,
  EXTERNAL_LIBRARIES: () => EXTERNAL_LIBRARIES,
  E_MODES: () => E_MODES,
  ORACLE: () => ORACLE,
  POOL: () => POOL,
  POOL_ADDRESSES_PROVIDER: () => POOL_ADDRESSES_PROVIDER,
  POOL_ADDRESSES_PROVIDER_REGISTRY: () => POOL_ADDRESSES_PROVIDER_REGISTRY,
  POOL_CONFIGURATOR: () => POOL_CONFIGURATOR,
  POOL_CONFIGURATOR_IMPL: () => POOL_CONFIGURATOR_IMPL,
  POOL_IMPL: () => POOL_IMPL,
  RISK_STEWARD: () => RISK_STEWARD,
  STATA_FACTORY: () => STATA_FACTORY,
  UI_INCENTIVE_DATA_PROVIDER: () => UI_INCENTIVE_DATA_PROVIDER,
  UI_POOL_DATA_PROVIDER: () => UI_POOL_DATA_PROVIDER,
  WALLET_BALANCE_PROVIDER: () => WALLET_BALANCE_PROVIDER,
  WETH_GATEWAY: () => WETH_GATEWAY
});
module.exports = __toCommonJS(AaveV3Sonic_exports);
var POOL_ADDRESSES_PROVIDER = "0x5C2e738F6E27bCE0F7558051Bf90605dD6176900";
var POOL = "0x5362dBb1e601abF3a4c14c22ffEdA64042E5eAA3";
var POOL_CONFIGURATOR = "0x50c70FEB95aBC1A92FC30b9aCc41Bd349E5dE2f0";
var ORACLE = "0xD63f7658C66B2934Bd234D79D06aEF5290734B30";
var ACL_ADMIN = "0x7b62461a3570c6AC8a9f8330421576e417B71EE7";
var ACL_MANAGER = "0x3a790a47c4d531FD333FAD24f70B0ccb521B3b5A";
var AAVE_PROTOCOL_DATA_PROVIDER = "0x306c124fFba5f2Bc0BcAf40D249cf19D492440b9";
var POOL_IMPL = "0x46Dcd5F4600319b02649Fd76B55aA6c1035CA478";
var POOL_CONFIGURATOR_IMPL = "0x5cC4f782cFe249286476A7eFfD9D7bd215768194";
var DEFAULT_INCENTIVES_CONTROLLER = "0x24bD6e9ca54F1737467DEf82dCA9702925B3Aa59";
var EMISSION_MANAGER = "0xd3A0A19Cdb7d1615f30988763BEA5f8feCc17a87";
var COLLECTOR = "0x1aB55bBdD5DF0782BBCf73553Af93BC6B29A286B";
var DEFAULT_A_TOKEN_IMPL_REV_1 = "0x91FC11136d5615575a0fC5981Ab5C0C54418E2C6";
var DEFAULT_VARIABLE_DEBT_TOKEN_IMPL_REV_1 = "0xcb85C501B3A5e9851850d66648d69B26A4c90942";
var CONFIG_ENGINE = "0xa119F84bC1b8083F5061E4cf53705cBf1065bA27";
var RISK_STEWARD = "0xAe13e4DA0952f0B8fE04E21df53716fCF799a923";
var POOL_ADDRESSES_PROVIDER_REGISTRY = "0x5E5D87858592d211fdb4503F09FB2B5cf805cB51";
var UI_INCENTIVE_DATA_PROVIDER = "0xeC859AED64e87004FA724892D5aB474901825B2D";
var UI_POOL_DATA_PROVIDER = "0x9005A69fE088680827f292e8aE885Be4BE1beb2f";
var WALLET_BALANCE_PROVIDER = "0x6637Ef9c6b41Fa62d69b7E88C5Dd9eF30BbEE945";
var WETH_GATEWAY = "0x061D8e131F26512348ee5FA42e2DF1bA9d6505E9";
var STATA_FACTORY = "0xFeeb6FE430B7523fEF2a38327241eE7153779535";
var DUST_BIN = "0x62aDD9C6B511085C67515ebB799b51c4cf2d786F";
var CHAIN_ID = 146;
var ASSETS = {
  WETH: {
    decimals: 18,
    id: 0,
    UNDERLYING: "0x50c42dEAcD8Fc9773493ED674b675bE577f2634b",
    A_TOKEN: "0xe18Ab82c81E7Eecff32B8A82B1b7d2d23F1EcE96",
    V_TOKEN: "0x07B1adFB7d5795Cf21baE8a77Eceb222F2FafBCE",
    INTEREST_RATE_STRATEGY: "0xdFf435BCcf782f11187D3a4454d96702eD78e092",
    ORACLE: "0x824364077993847f71293B24ccA8567c00c2de11",
    STATA_TOKEN: "0xeB5e9B0ae5bb60274786C747A1A2A798c11271E0"
  },
  USDCe: {
    decimals: 6,
    id: 1,
    UNDERLYING: "0x29219dd400f2Bf60E5a23d13Be72B486D4038894",
    A_TOKEN: "0x578Ee1ca3a8E1b54554Da1Bf7C583506C4CD11c6",
    V_TOKEN: "0x2273caBAd63b7D247A6b107E723c803fc49953A0",
    INTEREST_RATE_STRATEGY: "0xdFf435BCcf782f11187D3a4454d96702eD78e092",
    ORACLE: "0x7A8443a2a5D772db7f1E40DeFe32db485108F128",
    STATA_TOKEN: "0x6646248971427B80ce531bdD793e2Eb859347E55"
  },
  wS: {
    decimals: 18,
    id: 2,
    UNDERLYING: "0x039e2fB66102314Ce7b64Ce5Ce3E5183bc94aD38",
    A_TOKEN: "0x6C5E14A212c1C3e4Baf6f871ac9B1a969918c131",
    V_TOKEN: "0xF6089B790Fbf8F4812a79a31CFAbeB00B06BA7BD",
    INTEREST_RATE_STRATEGY: "0xdFf435BCcf782f11187D3a4454d96702eD78e092",
    ORACLE: "0xc76dFb89fF298145b417d221B2c747d84952e01d",
    STATA_TOKEN: "0x18B7B8695165290f2767BC63c36D3dFEa4C0F9bB"
  }
};
var E_MODES = {};
var EXTERNAL_LIBRARIES = {};
// Annotate the CommonJS export names for ESM import in node:
0 && (module.exports = {
  AAVE_PROTOCOL_DATA_PROVIDER,
  ACL_ADMIN,
  ACL_MANAGER,
  ASSETS,
  CHAIN_ID,
  COLLECTOR,
  CONFIG_ENGINE,
  DEFAULT_A_TOKEN_IMPL_REV_1,
  DEFAULT_INCENTIVES_CONTROLLER,
  DEFAULT_VARIABLE_DEBT_TOKEN_IMPL_REV_1,
  DUST_BIN,
  EMISSION_MANAGER,
  EXTERNAL_LIBRARIES,
  E_MODES,
  ORACLE,
  POOL,
  POOL_ADDRESSES_PROVIDER,
  POOL_ADDRESSES_PROVIDER_REGISTRY,
  POOL_CONFIGURATOR,
  POOL_CONFIGURATOR_IMPL,
  POOL_IMPL,
  RISK_STEWARD,
  STATA_FACTORY,
  UI_INCENTIVE_DATA_PROVIDER,
  UI_POOL_DATA_PROVIDER,
  WALLET_BALANCE_PROVIDER,
  WETH_GATEWAY
});
