// @bgd-labs/aave-address-book/dist/AaveV3Celo.js
"use strict";
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);

// src/ts/AaveV3Celo.ts
var AaveV3Celo_exports = {};
__export(AaveV3Celo_exports, {
  AAVE_PROTOCOL_DATA_PROVIDER: () => AAVE_PROTOCOL_DATA_PROVIDER,
  ACL_ADMIN: () => ACL_ADMIN,
  ACL_MANAGER: () => ACL_MANAGER,
  ASSETS: () => ASSETS,
  CHAIN_ID: () => CHAIN_ID,
  COLLECTOR: () => COLLECTOR,
  CONFIG_ENGINE: () => CONFIG_ENGINE,
  DEFAULT_A_TOKEN_IMPL_REV_1: () => DEFAULT_A_TOKEN_IMPL_REV_1,
  DEFAULT_INCENTIVES_CONTROLLER: () => DEFAULT_INCENTIVES_CONTROLLER,
  DEFAULT_VARIABLE_DEBT_TOKEN_IMPL_REV_1: () => DEFAULT_VARIABLE_DEBT_TOKEN_IMPL_REV_1,
  DUST_BIN: () => DUST_BIN,
  EMISSION_MANAGER: () => EMISSION_MANAGER,
  EXTERNAL_LIBRARIES: () => EXTERNAL_LIBRARIES,
  E_MODES: () => E_MODES,
  ORACLE: () => ORACLE,
  POOL: () => POOL,
  POOL_ADDRESSES_PROVIDER: () => POOL_ADDRESSES_PROVIDER,
  POOL_ADDRESSES_PROVIDER_REGISTRY: () => POOL_ADDRESSES_PROVIDER_REGISTRY,
  POOL_CONFIGURATOR: () => POOL_CONFIGURATOR,
  POOL_CONFIGURATOR_IMPL: () => POOL_CONFIGURATOR_IMPL,
  POOL_IMPL: () => POOL_IMPL,
  RISK_STEWARD: () => RISK_STEWARD,
  STATA_FACTORY: () => STATA_FACTORY,
  UI_INCENTIVE_DATA_PROVIDER: () => UI_INCENTIVE_DATA_PROVIDER,
  UI_POOL_DATA_PROVIDER: () => UI_POOL_DATA_PROVIDER,
  WALLET_BALANCE_PROVIDER: () => WALLET_BALANCE_PROVIDER
});
module.exports = __toCommonJS(AaveV3Celo_exports);
var POOL_ADDRESSES_PROVIDER = "0x9F7Cf9417D5251C59fE94fB9147feEe1aAd9Cea5";
var POOL = "0x3E59A31363E2ad014dcbc521c4a0d5757d9f3402";
var POOL_CONFIGURATOR = "0x7567E3434CC1BEf724AB595e6072367Ef4914691";
var ORACLE = "0x1e693D088ceFD1E95ba4c4a5F7EeA41a1Ec37e8b";
var ACL_ADMIN = "0x1dF462e2712496373A347f8ad10802a5E95f053D";
var ACL_MANAGER = "0x7a12dCfd73C1B4cddf294da4cFce75FcaBBa314C";
var AAVE_PROTOCOL_DATA_PROVIDER = "0x33b7d355613110b4E842f5f7057Ccd36fb4cee28";
var POOL_IMPL = "0x44D38b18a99E50E51B99F05C1f418Db26Ded315f";
var POOL_CONFIGURATOR_IMPL = "0x87f6224536d9BD1d4FE6052e06f9647B51843e33";
var DEFAULT_INCENTIVES_CONTROLLER = "0x4725A0FdbEB14A77964bC1C221eE3a7982263103";
var EMISSION_MANAGER = "0x9e04Cb339163b06068397d9b6af2dA78440954e0";
var COLLECTOR = "0xC959439207dA5341B74aDcdAC59016aa9Be7E9E7";
var DEFAULT_A_TOKEN_IMPL_REV_1 = "0x3F662AA079a9619F10BFF4848910195C52CB2fb4";
var DEFAULT_VARIABLE_DEBT_TOKEN_IMPL_REV_1 = "0x8Ff70eE5F8b607844A094a938E4DeD76aEbca5f0";
var CONFIG_ENGINE = "0xB25a5D144626a0D488e52AE717A051a2E9997076";
var POOL_ADDRESSES_PROVIDER_REGISTRY = "0xB4B6939D0804DCE8d0Ac3e59b2eD5C072829d56b";
var UI_INCENTIVE_DATA_PROVIDER = "0x3A1Ea5fB3098b9767E0e85D63E28A5c0EAa44099";
var UI_POOL_DATA_PROVIDER = "0xf07fFd12b119b921C4a2ce8d4A13C5d1E3000d6e";
var WALLET_BALANCE_PROVIDER = "0xB91DA65093d54a1a7cb0fe684860F568A5E57123";
var STATA_FACTORY = "0x2b33073B94243304bCb4dfFA6b624afA5BAA414D";
var RISK_STEWARD = "0x69a6CaF240698982c3Ac89E0A7C12E76bCEee4ef";
var DUST_BIN = "0x886b90e5d7DA22000f1652740D3EC51566AAbe40";
var CHAIN_ID = 42220;
var ASSETS = {};
var E_MODES = {};
var EXTERNAL_LIBRARIES = {};
// Annotate the CommonJS export names for ESM import in node:
0 && (module.exports = {
  AAVE_PROTOCOL_DATA_PROVIDER,
  ACL_ADMIN,
  ACL_MANAGER,
  ASSETS,
  CHAIN_ID,
  COLLECTOR,
  CONFIG_ENGINE,
  DEFAULT_A_TOKEN_IMPL_REV_1,
  DEFAULT_INCENTIVES_CONTROLLER,
  DEFAULT_VARIABLE_DEBT_TOKEN_IMPL_REV_1,
  DUST_BIN,
  EMISSION_MANAGER,
  EXTERNAL_LIBRARIES,
  E_MODES,
  ORACLE,
  POOL,
  POOL_ADDRESSES_PROVIDER,
  POOL_ADDRESSES_PROVIDER_REGISTRY,
  POOL_CONFIGURATOR,
  POOL_CONFIGURATOR_IMPL,
  POOL_IMPL,
  RISK_STEWARD,
  STATA_FACTORY,
  UI_INCENTIVE_DATA_PROVIDER,
  UI_POOL_DATA_PROVIDER,
  WALLET_BALANCE_PROVIDER
});
