// @bgd-labs/aave-address-book/dist/MiscArbitrum.js
"use strict";
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);

// src/ts/MiscArbitrum.ts
var MiscArbitrum_exports = {};
__export(MiscArbitrum_exports, {
  AAVE_ARB_ETH_BRIDGE: () => AAVE_ARB_ETH_BRIDGE,
  AAVE_CL_ROBOT_OPERATOR: () => AAVE_CL_ROBOT_OPERATOR,
  CHAIN_ID: () => CHAIN_ID,
  CREATE_3_FACTORY: () => CREATE_3_FACTORY,
  LEGACY_BRIDGE_EXECUTOR: () => LEGACY_BRIDGE_EXECUTOR,
  PROTOCOL_GUARDIAN: () => PROTOCOL_GUARDIAN,
  PROXY_ADMIN: () => PROXY_ADMIN,
  TRANSPARENT_PROXY_FACTORY: () => TRANSPARENT_PROXY_FACTORY,
  rETH_ETH_AGGREGATOR: () => rETH_ETH_AGGREGATOR,
  wstETH_stETH_AGGREGATOR: () => wstETH_stETH_AGGREGATOR
});
module.exports = __toCommonJS(MiscArbitrum_exports);
var TRANSPARENT_PROXY_FACTORY = "0xEB0682d148e874553008730f0686ea89db7DA412";
var PROXY_ADMIN = "0xD3cF979e676265e4f6379749DECe4708B9A22476";
var CREATE_3_FACTORY = "0x3b56998Ec06477704622ca8e2eA1b4db134cec32";
var AAVE_CL_ROBOT_OPERATOR = "0xaa944aD95e51CB83C1f35FAEEDfC7d2c31B0BB4d";
var PROTOCOL_GUARDIAN = "0xCb45E82419baeBCC9bA8b1e5c7858e48A3B26Ea6";
var wstETH_stETH_AGGREGATOR = "0xB1552C5e96B312d0Bf8b554186F846C40614a540";
var rETH_ETH_AGGREGATOR = "0xF3272CAfe65b190e76caAF483db13424a3e23dD2";
var LEGACY_BRIDGE_EXECUTOR = "0x7d9103572bE58FfE99dc390E8246f02dcAe6f611";
var AAVE_ARB_ETH_BRIDGE = "0x0335ffa9af5CE05590d6C9A75B645470e07744a9";
var CHAIN_ID = 42161;
// Annotate the CommonJS export names for ESM import in node:
0 && (module.exports = {
  AAVE_ARB_ETH_BRIDGE,
  AAVE_CL_ROBOT_OPERATOR,
  CHAIN_ID,
  CREATE_3_FACTORY,
  LEGACY_BRIDGE_EXECUTOR,
  PROTOCOL_GUARDIAN,
  PROXY_ADMIN,
  TRANSPARENT_PROXY_FACTORY,
  rETH_ETH_AGGREGATOR,
  wstETH_stETH_AGGREGATOR
});
