// @bgd-labs/aave-address-book/dist/MiscEthereum.mjs
// src/ts/MiscEthereum.ts
var AAVE_ECOSYSTEM_RESERVE_CONTROLLER = "0x3d569673dAa0575c936c7c67c4E6AedA69CC630C";
var ECOSYSTEM_RESERVE = "0x25F2226B597E8F9514B3F68F00f494cF4f286491";
var PROXY_ADMIN_LONG = "0x86C3FfeE349A7cFf7cA88C449717B1b133bfb517";
var AAVE_SWAPPER = "0x3ea64b1C0194524b48F9118462C8E9cd61a243c7";
var AAVE_POL_ETH_BRIDGE = "0x1C2BA5b8ab8e795fF44387ba6d251fa65AD20b36";
var TRANSPARENT_PROXY_FACTORY = "0xEB0682d148e874553008730f0686ea89db7DA412";
var PROXY_ADMIN = "0xD3cF979e676265e4f6379749DECe4708B9A22476";
var CREATE_3_FACTORY = "0xcc3C54B95f3f1867A43009B80ed4DD930E3cE2F7";
var AAVE_CL_ROBOT_OPERATOR = "0x1cDF8879eC8bE012bA959EB515b11008E0cb6323";
var PROTOCOL_GUARDIAN = "0x2CFe3ec4d5a6811f4B8067F0DE7e47DfA938Aa30";
var AAVE_MERKLE_DISTRIBUTOR = "0xa88c6D90eAe942291325f9ae3c66f3563B93FE10";
var sDAI_POT = "0x197E90f9FAD81970bA7976f33CbD77088E5D7cf7";
var stEUR = "0x004626A008B1aCdC4c74ab51644093b155e59A23";
var agEUR_EUR_AGGREGATOR = "0xb4d5289C58CE36080b0748B47F859D8F50dFAACb";
var EUR_USD_AGGREGATOR = "0xb49f677943BC038e9857d61E7d053CaA2C1734C1";
var weETH_RATIO_PROVIDER = "0xCd5fE23C85820F7B72D0926FC9b05b43E359b7ee";
var CHAIN_ID = 1;
export {
  AAVE_CL_ROBOT_OPERATOR,
  AAVE_ECOSYSTEM_RESERVE_CONTROLLER,
  AAVE_MERKLE_DISTRIBUTOR,
  AAVE_POL_ETH_BRIDGE,
  AAVE_SWAPPER,
  CHAIN_ID,
  CREATE_3_FACTORY,
  ECOSYSTEM_RESERVE,
  EUR_USD_AGGREGATOR,
  PROTOCOL_GUARDIAN,
  PROXY_ADMIN,
  PROXY_ADMIN_LONG,
  TRANSPARENT_PROXY_FACTORY,
  agEUR_EUR_AGGREGATOR,
  sDAI_POT,
  stEUR,
  weETH_RATIO_PROVIDER
};
