// @hashflow/sdk/dist/legacy/v0_1/contracts/protocol/factories/IHashflowTracker__factory.js
"use strict";
/* Autogenerated file. Do not edit manually. */
/* tslint:disable */
/* eslint-disable */
Object.defineProperty(exports, "__esModule", { value: true });
exports.IHashflowTracker__factory = void 0;
const ethers_1 = require("ethers");
const _abi = [
    {
        anonymous: false,
        inputs: [
            {
                indexed: false,
                internalType: "address",
                name: "fromPool",
                type: "address",
            },
            {
                indexed: false,
                internalType: "address",
                name: "toPool",
                type: "address",
            },
            {
                indexed: false,
                internalType: "address",
                name: "token",
                type: "address",
            },
            {
                indexed: false,
                internalType: "enum IHashflowTracker.HTokenType",
                name: "hTokenType",
                type: "uint8",
            },
            {
                indexed: false,
                internalType: "address",
                name: "hToken",
                type: "address",
            },
        ],
        name: "HTokenMigrated",
        type: "event",
    },
    {
        anonymous: false,
        inputs: [
            {
                indexed: false,
                internalType: "address",
                name: "pool",
                type: "address",
            },
            {
                indexed: false,
                internalType: "address",
                name: "token",
                type: "address",
            },
            {
                indexed: false,
                internalType: "enum IHashflowTracker.HTokenType",
                name: "hTokenType",
                type: "uint8",
            },
            {
                indexed: false,
                internalType: "address",
                name: "hToken",
                type: "address",
            },
        ],
        name: "HTokenRegistered",
        type: "event",
    },
    {
        anonymous: false,
        inputs: [
            {
                indexed: false,
                internalType: "address",
                name: "governance",
                type: "address",
            },
            {
                indexed: false,
                internalType: "address",
                name: "prevGovernance",
                type: "address",
            },
        ],
        name: "UpdateGovernance",
        type: "event",
    },
    {
        anonymous: false,
        inputs: [
            {
                indexed: false,
                internalType: "address",
                name: "pool",
                type: "address",
            },
            {
                indexed: false,
                internalType: "bool",
                name: "status",
                type: "bool",
            },
        ],
        name: "UpdatePoolAuthorization",
        type: "event",
    },
    {
        inputs: [
            {
                internalType: "string",
                name: "_name",
                type: "string",
            },
            {
                internalType: "string",
                name: "_symbol",
                type: "string",
            },
            {
                internalType: "enum IHashflowTracker.HTokenType",
                name: "hTokenType",
                type: "uint8",
            },
            {
                internalType: "address",
                name: "token",
                type: "address",
            },
        ],
        name: "deployHToken",
        outputs: [
            {
                internalType: "address",
                name: "",
                type: "address",
            },
        ],
        stateMutability: "nonpayable",
        type: "function",
    },
    {
        inputs: [],
        name: "getGovernance",
        outputs: [
            {
                internalType: "address",
                name: "",
                type: "address",
            },
        ],
        stateMutability: "view",
        type: "function",
    },
    {
        inputs: [
            {
                internalType: "address",
                name: "pool",
                type: "address",
            },
            {
                internalType: "enum IHashflowTracker.HTokenType",
                name: "hTokenType",
                type: "uint8",
            },
            {
                internalType: "address",
                name: "token",
                type: "address",
            },
        ],
        name: "getHToken",
        outputs: [
            {
                internalType: "address",
                name: "",
                type: "address",
            },
        ],
        stateMutability: "view",
        type: "function",
    },
    {
        inputs: [
            {
                internalType: "address",
                name: "hToken",
                type: "address",
            },
        ],
        name: "getHTokenPool",
        outputs: [
            {
                internalType: "address",
                name: "",
                type: "address",
            },
        ],
        stateMutability: "view",
        type: "function",
    },
    {
        inputs: [
            {
                internalType: "address",
                name: "hToken",
                type: "address",
            },
        ],
        name: "getHTokenToken",
        outputs: [
            {
                internalType: "address",
                name: "",
                type: "address",
            },
        ],
        stateMutability: "view",
        type: "function",
    },
    {
        inputs: [
            {
                internalType: "address",
                name: "hToken",
                type: "address",
            },
        ],
        name: "getHTokenType",
        outputs: [
            {
                internalType: "enum IHashflowTracker.HTokenType",
                name: "",
                type: "uint8",
            },
        ],
        stateMutability: "view",
        type: "function",
    },
    {
        inputs: [
            {
                internalType: "address",
                name: "pool",
                type: "address",
            },
        ],
        name: "getPoolAuthorization",
        outputs: [
            {
                internalType: "bool",
                name: "",
                type: "bool",
            },
        ],
        stateMutability: "view",
        type: "function",
    },
    {
        inputs: [
            {
                internalType: "address",
                name: "fromPool",
                type: "address",
            },
            {
                internalType: "enum IHashflowTracker.HTokenType",
                name: "hTokenType",
                type: "uint8",
            },
            {
                internalType: "address",
                name: "token",
                type: "address",
            },
        ],
        name: "migrateHToken",
        outputs: [],
        stateMutability: "nonpayable",
        type: "function",
    },
    {
        inputs: [
            {
                internalType: "address",
                name: "token",
                type: "address",
            },
            {
                internalType: "enum IHashflowTracker.HTokenType",
                name: "hTokenType",
                type: "uint8",
            },
            {
                internalType: "address",
                name: "hToken",
                type: "address",
            },
        ],
        name: "registerHToken",
        outputs: [],
        stateMutability: "nonpayable",
        type: "function",
    },
    {
        inputs: [
            {
                internalType: "address",
                name: "pool",
                type: "address",
            },
        ],
        name: "registerPool",
        outputs: [],
        stateMutability: "nonpayable",
        type: "function",
    },
    {
        inputs: [
            {
                internalType: "address",
                name: "governance",
                type: "address",
            },
        ],
        name: "updateGovernance",
        outputs: [],
        stateMutability: "nonpayable",
        type: "function",
    },
    {
        inputs: [
            {
                internalType: "address",
                name: "pool",
                type: "address",
            },
            {
                internalType: "bool",
                name: "status",
                type: "bool",
            },
        ],
        name: "updatePoolAuthorization",
        outputs: [],
        stateMutability: "nonpayable",
        type: "function",
    },
    {
        inputs: [
            {
                internalType: "bytes32",
                name: "txid",
                type: "bytes32",
            },
        ],
        name: "validateTxid",
        outputs: [],
        stateMutability: "nonpayable",
        type: "function",
    },
];
class IHashflowTracker__factory {
    static createInterface() {
        return new ethers_1.Interface(_abi);
    }
    static connect(address, runner) {
        return new ethers_1.Contract(address, _abi, runner);
    }
}
exports.IHashflowTracker__factory = IHashflowTracker__factory;
IHashflowTracker__factory.abi = _abi;
