// @hashflow/protocol/dist/typechain-types/factories/contracts/testing/index.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.WETH9__factory = exports.USDT__factory = exports.USDC__factory = exports.TestToken2__factory = exports.TestToken1__factory = exports.LZEndpointMock__factory = exports.DummyMKR__factory = void 0;
/* Autogenerated file. Do not edit manually. */
/* tslint:disable */
/* eslint-disable */
var DummyMKR__factory_1 = require("./DummyMKR__factory");
Object.defineProperty(exports, "DummyMKR__factory", { enumerable: true, get: function () { return DummyMKR__factory_1.DummyMKR__factory; } });
var LZEndpointMock__factory_1 = require("./LZEndpointMock__factory");
Object.defineProperty(exports, "LZEndpointMock__factory", { enumerable: true, get: function () { return LZEndpointMock__factory_1.LZEndpointMock__factory; } });
var TestToken1__factory_1 = require("./TestToken1__factory");
Object.defineProperty(exports, "TestToken1__factory", { enumerable: true, get: function () { return TestToken1__factory_1.TestToken1__factory; } });
var TestToken2__factory_1 = require("./TestToken2__factory");
Object.defineProperty(exports, "TestToken2__factory", { enumerable: true, get: function () { return TestToken2__factory_1.TestToken2__factory; } });
var USDC__factory_1 = require("./USDC__factory");
Object.defineProperty(exports, "USDC__factory", { enumerable: true, get: function () { return USDC__factory_1.USDC__factory; } });
var USDT__factory_1 = require("./USDT__factory");
Object.defineProperty(exports, "USDT__factory", { enumerable: true, get: function () { return USDT__factory_1.USDT__factory; } });
var WETH9__factory_1 = require("./WETH9__factory");
Object.defineProperty(exports, "WETH9__factory", { enumerable: true, get: function () { return WETH9__factory_1.WETH9__factory; } });
