// @hashflow/sdk/dist/contracts/protocol/factories/IHashflowFactory__factory.js
"use strict";
/* Autogenerated file. Do not edit manually. */
/* tslint:disable */
/* eslint-disable */
Object.defineProperty(exports, "__esModule", { value: true });
exports.IHashflowFactory__factory = void 0;
const ethers_1 = require("ethers");
const _abi = [
    {
        anonymous: false,
        inputs: [
            {
                indexed: false,
                internalType: "address",
                name: "pool",
                type: "address",
            },
            {
                indexed: false,
                internalType: "address",
                name: "operations",
                type: "address",
            },
        ],
        name: "CreatePool",
        type: "event",
    },
    {
        anonymous: false,
        inputs: [
            {
                indexed: false,
                internalType: "address",
                name: "poolCreator",
                type: "address",
            },
            {
                indexed: false,
                internalType: "bool",
                name: "authorizationStatus",
                type: "bool",
            },
        ],
        name: "UpdatePoolCreatorAuthorization",
        type: "event",
    },
    {
        anonymous: false,
        inputs: [
            {
                indexed: false,
                internalType: "address",
                name: "poolImpl",
                type: "address",
            },
            {
                indexed: false,
                internalType: "address",
                name: "prevPoolImpl",
                type: "address",
            },
        ],
        name: "UpdatePoolImplementation",
        type: "event",
    },
    {
        inputs: [
            {
                internalType: "address",
                name: "poolCreator",
                type: "address",
            },
        ],
        name: "allowedPoolCreators",
        outputs: [
            {
                internalType: "bool",
                name: "",
                type: "bool",
            },
        ],
        stateMutability: "view",
        type: "function",
    },
    {
        inputs: [
            {
                internalType: "string",
                name: "name",
                type: "string",
            },
            {
                internalType: "address",
                name: "signer",
                type: "address",
            },
        ],
        name: "createPool",
        outputs: [],
        stateMutability: "nonpayable",
        type: "function",
    },
    {
        inputs: [
            {
                internalType: "address",
                name: "router",
                type: "address",
            },
        ],
        name: "initialize",
        outputs: [],
        stateMutability: "nonpayable",
        type: "function",
    },
    {
        inputs: [],
        name: "router",
        outputs: [
            {
                internalType: "address",
                name: "",
                type: "address",
            },
        ],
        stateMutability: "view",
        type: "function",
    },
    {
        inputs: [
            {
                internalType: "address",
                name: "poolCreator",
                type: "address",
            },
            {
                internalType: "bool",
                name: "status",
                type: "bool",
            },
        ],
        name: "updatePoolCreatorAuthorization",
        outputs: [],
        stateMutability: "nonpayable",
        type: "function",
    },
    {
        inputs: [
            {
                internalType: "address",
                name: "poolImpl",
                type: "address",
            },
        ],
        name: "updatePoolImpl",
        outputs: [],
        stateMutability: "nonpayable",
        type: "function",
    },
];
class IHashflowFactory__factory {
    static createInterface() {
        return new ethers_1.Interface(_abi);
    }
    static connect(address, runner) {
        return new ethers_1.Contract(address, _abi, runner);
    }
}
exports.IHashflowFactory__factory = IHashflowFactory__factory;
IHashflowFactory__factory.abi = _abi;
