// @hashflow/protocol/dist/hardhat.config.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    Object.defineProperty(o, k2, { enumerable: true, get: function() { return m[k]; } });
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || function (mod) {
    if (mod && mod.__esModule) return mod;
    var result = {};
    if (mod != null) for (var k in mod) if (k !== "default" && Object.prototype.hasOwnProperty.call(mod, k)) __createBinding(result, mod, k);
    __setModuleDefault(result, mod);
    return result;
};
Object.defineProperty(exports, "__esModule", { value: true });
const dotenv = __importStar(require("dotenv"));
require("@nomiclabs/hardhat-etherscan");
require("@typechain/hardhat");
require("@matterlabs/hardhat-zksync-deploy");
require("@matterlabs/hardhat-zksync-solc");
require("hardhat-gas-reporter");
require("hardhat-contract-sizer");
require("./tasks/deploy");
require("./tasks/initialize");
require("./tasks/wormhole");
dotenv.config();
const networksUserConfig = {
    ethereum: {
        chainId: 1,
        zksync: false,
        url: process.env.ETHEREUM_URL || '',
        accounts: process.env.PRIVATE_KEY !== undefined ? [process.env.PRIVATE_KEY] : [],
    },
    polygon: {
        chainId: 137,
        zksync: false,
        url: process.env.POLYGON_URL || '',
        accounts: process.env.PRIVATE_KEY !== undefined ? [process.env.PRIVATE_KEY] : [],
    },
    goerli: {
        chainId: 5,
        zksync: false,
        url: process.env.GOERLI_URL || '',
        accounts: process.env.PRIVATE_KEY !== undefined ? [process.env.PRIVATE_KEY] : [],
    },
    'zksync-testnet': {
        chainId: 280,
        zksync: true,
        url: process.env.ZKSYNC_TESTNET_URL || '',
        accounts: process.env.PRIVATE_KEY !== undefined ? [process.env.PRIVATE_KEY] : [],
    },
    mumbai: {
        chainId: 80001,
        zksync: false,
        url: process.env.MUMBAI_URL || '',
        accounts: process.env.PRIVATE_KEY !== undefined ? [process.env.PRIVATE_KEY] : [],
    },
    'bnb-testnet': {
        chainId: 97,
        zksync: false,
        url: process.env.BNB_TESTNET_URL || '',
        accounts: process.env.PRIVATE_KEY !== undefined ? [process.env.PRIVATE_KEY] : [],
    },
    bnb: {
        chainId: 56,
        zksync: false,
        url: process.env.BNB_URL || '',
        accounts: process.env.PRIVATE_KEY !== undefined ? [process.env.PRIVATE_KEY] : [],
    },
    arbitrum: {
        chainId: 42161,
        zksync: false,
        url: process.env.ARBITRUM_URL || '',
        accounts: process.env.PRIVATE_KEY !== undefined ? [process.env.PRIVATE_KEY] : [],
    },
    avalanche: {
        chainId: 43114,
        zksync: false,
        url: process.env.AVALANCHE_URL || '',
        accounts: process.env.PRIVATE_KEY !== undefined ? [process.env.PRIVATE_KEY] : [],
    },
    optimism: {
        chainId: 10,
        zksync: false,
        url: process.env.OPTIMISM_URL || '',
        accounts: process.env.PRIVATE_KEY !== undefined ? [process.env.PRIVATE_KEY] : [],
    },
    // These networks are used to test Wormhole x-calls locally.
    // They are run in two different Ganache EVM processes
    wormhole1: {
        chainId: 1337,
        zksync: false,
        url: 'http://localhost:8545',
        accounts: [
            '0x4f3edf983ac636a65a842ce7c78d9aa706d3b113bce9c46f30d7d21715b23b1d',
            '0xac0974bec39a17e36ba4a6b4d238ff944bacb478cbed5efcae784d7bf4f2ff80',
            '0x5de4111afa1a4b94908f83103eb1f1706367c2e68ca870fc3fb9a804cdab365a', // Trader
        ],
    },
    wormhole2: {
        chainId: 1337,
        zksync: false,
        url: 'http://localhost:8546',
        accounts: [
            '0x4f3edf983ac636a65a842ce7c78d9aa706d3b113bce9c46f30d7d21715b23b1d',
            '0xac0974bec39a17e36ba4a6b4d238ff944bacb478cbed5efcae784d7bf4f2ff80',
            '0x5de4111afa1a4b94908f83103eb1f1706367c2e68ca870fc3fb9a804cdab365a', // Trader
        ],
    },
};
// You need to export an object to set up your config
// Go to https://hardhat.org/config/ to learn more
const config = {
    solidity: {
        version: '0.8.13',
        settings: {
            optimizer: {
                enabled: true,
                runs: 200,
            },
        },
    },
    networks: networksUserConfig,
    gasReporter: {
        enabled: process.env.REPORT_GAS !== undefined,
        currency: 'USD',
    },
    etherscan: {
        apiKey: process.env.ETHERSCAN_API_KEY,
    },
    contractSizer: {
        alphaSort: true,
        disambiguatePaths: false,
        runOnCompile: true,
        strict: true,
    },
    zksolc: {
        version: '1.1.5',
        compilerSource: 'docker',
        settings: {
            experimental: {
                dockerImage: 'matterlabs/zksolc',
                tag: 'v1.1.6',
            },
        },
    },
    zkSyncDeploy: {
        zkSyncNetwork: 'https://zksync2-testnet.zksync.dev',
        ethNetwork: 'goerli', // Can also be the RPC URL of the network (e.g. `https://goerli.infura.io/v3/<API_KEY>`)
    },
};
extendEnvironment((hre) => {
    hre.config.contractSizer.runOnCompile =
        hre.hardhatArguments.network !== 'zksync-testnet';
});
exports.default = config;
