// @hashflow/sdk/dist/legacy/v0_1/Hashflow.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.Hashflow = void 0;
const Contracts_1 = require("./Contracts");
const Optimism_1 = require("./modules/Optimism");
const Pool_1 = require("./modules/Pool");
const Token_1 = require("./modules/Token");
const types_1 = require("./types");
class Hashflow {
    constructor({ networkId = types_1.Networks.MAINNET, signer, provider, }) {
        this.contracts = new Contracts_1.Contracts(networkId, signer, provider);
        this.pool = new Pool_1.Pool(this.contracts);
        this.token = new Token_1.Token(this.contracts);
        this.optimism = new Optimism_1.Optimism(this.contracts);
        if (provider) {
            this.provider = provider;
        }
        else if (signer) {
            this.signer = signer;
            if (!signer.provider) {
                throw new Error(`Supplied Signer does not have a Provider.`);
            }
            this.provider = signer.provider;
        }
        else {
            throw new Error('Missing provider');
        }
    }
    getBlockNumber() {
        return this.provider.getBlockNumber();
    }
}
exports.Hashflow = Hashflow;
